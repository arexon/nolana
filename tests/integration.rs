use std::{fs, path::Path};

use insta::Settings;
use nolana::{Codegen, CodegenOptions, MolangTransformer, Parser, semantic::SemanticChecker};

fn with_settings(f: impl FnOnce()) {
    let mut settings = Settings::clone_current();
    settings.set_omit_expression(true);
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(f)
}

fn read_and_parse(path: &Path) -> String {
    let source = fs::read_to_string(path).unwrap();
    let result = Parser::new(&source).parse();
    format!("{result:#?}")
}

fn read_and_codegen(path: &Path) -> String {
    let source = fs::read_to_string(path).unwrap();
    let result = Parser::new(&source).parse();
    assert!(result.errors.is_empty());
    assert!(!result.panicked);
    Codegen::default().build(&result.program)
}

fn read_and_semantic(path: &Path) -> String {
    let source = fs::read_to_string(path).unwrap();
    let mut result = Parser::new(&source).parse();
    let diagnostics = SemanticChecker::default().check(&mut result.program);
    format!("{diagnostics:#?}")
}

fn read_and_transform(path: &Path) -> String {
    let source = fs::read_to_string(path).unwrap();
    let mut result = Parser::new(&source).parse();
    MolangTransformer::default().transform(&mut result.program);
    Codegen::default().with_options(CodegenOptions { minify: false }).build(&result.program)
}

#[test]
fn test_parser() {
    with_settings(|| {
        insta::glob!("parser/*.nolana", |path| {
            insta::assert_snapshot!(read_and_parse(path));
        });
    });
}

#[test]
fn test_codegen() {
    with_settings(|| {
        insta::glob!("codegen/*.nolana", |path| {
            insta::assert_snapshot!(read_and_codegen(path));
        });
    });
}

#[test]
fn test_semantic() {
    with_settings(|| {
        insta::glob!("semantic/*.nolana", |path| {
            insta::assert_snapshot!(read_and_semantic(path));
        });
    });
}

#[test]
fn test_transformer() {
    with_settings(|| {
        insta::glob!("transformer/*.nolana", |path| {
            insta::assert_snapshot!(read_and_transform(path));
        });
    });
}
