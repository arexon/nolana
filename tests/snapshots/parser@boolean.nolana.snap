---
source: tests/integration.rs
input_file: tests/parser/boolean.nolana
---
ParseResult {
    program: Program {
        span: Span {
            start: 0,
            end: 12,
        },
        source: "false; true;\n",
        body: Complex(
            [
                Expression(
                    BooleanLiteral(
                        BooleanLiteral {
                            span: Span {
                                start: 0,
                                end: 5,
                            },
                            value: false,
                        },
                    ),
                ),
                Expression(
                    BooleanLiteral(
                        BooleanLiteral {
                            span: Span {
                                start: 7,
                                end: 11,
                            },
                            value: true,
                        },
                    ),
                ),
            ],
        ),
    },
    errors: [],
    panicked: false,
}
